core/src/main/java/feast/core/service/JobService.java

```diff
@@ -0,0 +1,223 @@
+
+        List<String> jobIds =
+            jobs.stream()
+                .map(
+                    job -> {
+                      return job.getId();
+                    })
+                .collect(Collectors.toList());
+        matchingJobIds = this.mergeResults(matchingJobIds, jobIds);
+      }
+    }
+
+    // convert matching job models to ingestion job protos
+    List<IngestionJobProto.IngestionJob> ingestJobs = new ArrayList<>();
+    for (String jobId : matchingJobIds) {
+      Job job = this.jobRepository.findById(jobId).get();
+      ingestJobs.add(job.toIngestionProto());
+    }
+
+    // pack jobs into response
+    return ListIngestionJobsResponse.newBuilder().addAllJobs(ingestJobs).build();
+  }
+
+  // TODO: restart ingestion job
+
+  /**
+   * Stops (Aborts) the ingestion job matching the given request. Does nothing if the target job to
+   * be stopped is already stopped or stopping
+   *
+   * @param request stop ingestion job request specifying which job to stop
+   * @throws NoSuchElementException when stop job request requests to stop a nonexistent job.
+   * @throws UnsupportedOperationException when job to be stopped is in an unknown status
+   * @throws InvalidProtocolBufferException on error when constructing response protobuf
+   */
+  public StopIngestionJobResponse stopJob(StopIngestionJobRequest request)
+      throws InvalidProtocolBufferException {
+    // check job exists
+    Optional<Job> getJob = this.jobRepository.findById(request.getId());
+    if (getJob.isEmpty()) {
+      throw new NoSuchElementException(
+          "Attempted to stop nonexistent job with id: " + getJob.get().getId());
+    }
+
+    // check job status is valid for stopping
+    Job job = getJob.get();
+    JobStatus status = job.getStatus();
+    if (status.equals(JobStatus.ABORTED)
+        || status.equals(JobStatus.ABORTING)
+        || status.equals(JobStatus.SUSPENDED)
+        || status.equals(JobStatus.SUSPENDING)
+        || status.equals(JobStatus.COMPLETED)
+        || status.equals(JobStatus.ERROR)) {
+      // do nothing - job is already stopped or stopping
+      return StopIngestionJobResponse.newBuilder().build();
+    } else if (status.equals(JobStatus.UNKNOWN)) {
+      throw new UnsupportedOperationException(
+          "Stopping a job with an unknown status is unsupported");
+    }
+
+    // stop job with job manager
+    JobManager jobManager = this.jobManagers.get(job.getRunner());
+    jobManager.abortJob(job.getExtId());
+
+    return StopIngestionJobResponse.newBuilder().build();
+  }
+
+  /* Private Utility Methods */
+  /**
+   * Finds &amp; returns featuresets matching the given feature set refererence
+   *
+   * @param fsReference FeatureSetReference that specifies which featuresets to match
+   * @throws UnsupportedOperationException fsReference given is unsupported.
+   * @return Returns a list of matching featuresets
+   */
+  private List<FeatureSet> findFeatureSets(FeatureSetReference fsReference)
```

Comment:
This should almost certainly be externalized. Can you have a look in the code base to see where else it occurs and see if we can reuse or generalize this?

Reply:
I moved the this method to `SpecService` as `matchFeatureSets()`.